### PR TITLE
Improve user search experience #6872

### LINKED ISSUE #6872: [OTHER] Improve user search experience
### What would you like to share?

These are the issues relating the user search experience on the search page that need to improved/fixed:
- Search keywords including a name along with a tag separated by a comma ("OpenSource,Eddie Jaoude") results in not found error.
- If there is an extra comma or an extra space, it should be ignored otherwise it has a bad user search experience.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by itsdevdeepak**
## Fixes Issue

Closes #6872 

## Changes proposed

- [x] Updated the `filterData` method to to filler users that matches all searched keywords (comma separated values). 
- [x] Updated the `search` method to use `cleanInput` and  use `searchTagNameInInput` to search tag name form search input.
- [x] Ignores the extra spaces and commas in search input. 
- [x] Moved the `useEffect` hook to top of the components according to the rule of hooks.
- [x] Moved the `result` to the bottom of the hooks.

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

[video preview](https://imgur.com/a/VdW8o3S)

## Note to reviewers

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/7392"></a>

### CODE REVIEW

**Review by Vishrut19 — APPROVED** [review]
Looks good to me! 👍

File: pages/search.js [hunk 1]
```diff
@@ -47,36 +47,63 @@ export default function Search({ data: { users, tags, randUsers } }) {
   const router = useRouter();
   const { username, keyword } = router.query;
   const [notFound, setNotFound] = useState();
-
   const [filteredUsers, setFilteredUsers] = useState(randUsers);
   const [inputValue, setInputValue] = useState(username || keyword || "");
+  
   let results = [];
-
+
```
Comment (eddiejaoude) [inline pages/search.js:54]: Blank spaces not required, you can setup Prettier to format this correctly on save in VScode

```suggestion

```

File: pages/search.js [hunk 2]
```diff
@@ -47,36 +47,62 @@ export default function Search({ data: { users, tags, randUsers } }) {
   const router = useRouter();
   const { username, keyword } = router.query;
   const [notFound, setNotFound] = useState();
-
   const [filteredUsers, setFilteredUsers] = useState(randUsers);
   const [inputValue, setInputValue] = useState(username || keyword || "");
+
```
Comment (eddiejaoude) [inline pages/search.js:52]:
```suggestion

```

**Review by eddiejaoude — CHANGES_REQUESTED** [review]
Looks good 👍 I am not sure if the results are always correct though

Searching for `Eddie` on production...

Screenshot 2023-06-09 at 11 37 16

Searching on your PR... (I think 2 results should not be there as they are not in their `username`, `name` or `tags`?)

Screenshot 2023-06-09 at 11 36 44

**Comment by itsdevdeepak** [comment]
> Looks good 👍 I am not sure if the results are always correct though
> 
> Searching for `Eddie` on production...
> Screenshot 2023-06-09 at 11 37 16
> 
> Searching on your PR... (I think 2 results should not be there as they are not in their `username`, `name` or `tags`?)
> Screenshot 2023-06-09 at 11 36 44

i think its because they are using `eddiehub`  tag in their profile

**Comment by eddiejaoude** [comment]
>> i think its because they are using `eddiehub` tag in their profile.
> 
> should i change the code to only search for tags if they are complete matched ?

Oh my bad, let me double check their tags - if they are then the results are correct

**Comment by eddiejaoude** [comment]
Yes you are correct 👍 , they have it in their tags https://github.com/EddieHubCommunity/LinkFree/blob/main/data/LinuxGamer.json#L4

**Review by eddiejaoude — APPROVED** [review]
Thank you 👍

**Pull request merged**